### PR TITLE
Enable completions for blank between comma and something

### PR DESCRIPTION
**Pull request by matt9ucci**
# PR Summary

Tab completions for a blank between a comma and something (e.g. parameter) does not work.

For example, the No. 1 in the following cases works well, but the No. 2 does not:

No. | o/x | Command and \<tab\>                   | `lastAst`     | `lastAst is`
--: | :-: | :------------------------------------ | :------------ | :----------------
  1 | ⭕ | `Get-Command -Type Alias,<tab>`       | {Alias,}      | [ErrorExpressionAst]
  2 | ❌ | `Get-Command -Type Alias,<tab> -All`  | {Alias, -All} | [ArrayLiteralAst]

That is because the tab completion is triggered if the `lastAst` is `ErrorExpressionAst`. Not triggered if `ArrayLiteralAst`. There are same issues other than the `Get-Command`, for example, `dir` (`Get-ChildItem`), `Get-Alias`, `Get-Module`, etc.

I have fixed the condition of triggering the tab completion. It is triggered if the `lastAst` is `ErrorExpressionAst` or `ArrayLiteralAst`.

## PR Context

It helps us to modify such an array, especially when we use Alt+a key (SelectCommandArguments).

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
- **User-facing changes**
    - [x] Not Applicable
- **Testing - New and feature**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [x] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.

### CODE REVIEW

**Comment by daxian-dbw** [comment]
@MartinGC94, can you please take a quick look at this PR and see if there could be any unwanted side effects after adding `|| lastAst is ArrayLiteralAst` in the check?

**Comment by daxian-dbw** [comment]
@MartinGC94 Thanks for catching it! Just tried and found that `Get-Command -Type Alias, <Tab> -All` doesn't work too as of today, so I guess the change fixed that but not `-Type Alias,<tab> -All`?

**Comment by MartinGC94** [comment]
No this PR doesn't seem to fix anything. `Get-Command -Type Alias, <Tab> -All` doesn't work because no code was changed to fix it from its current broken state. `-Type Alias,<tab> -All` doesn't work because of the previously mentioned null exception further down the method chain.

**Comment by matt9ucci** [comment]
@daxian-dbw @MartinGC94 Thank you for your reply and I am sorry for my mistake. I have tested it again.

It works for the following commands:

* `Get-ChildItem -Path .\a.txt,<TAB> -Recurse`
* `Get-Alias -Name gal,<TAB> -Verbose`
* `Get-Module -Name PSReadLine,<TAB> -All`

But it does not work, as you say, for `Get-Command -Type Alias,<TAB> -All`.

Why? I think it is because the `-Type` parameter (alias for `-CommandType`) accepts `[System.Management.Automation.CommandTypes]` enum flags. I have noticed that all the above success cases are `[string[]]` parameter.

Actually, my motive in the first place is to fix `Trace-Command -Name CmdletProviderClasses,<TAB> -Expression {}`. The `-Name` is also `[string[]]`. But I mistook the PR could solve more wide range of problems including `Get-Command`.

Additionally, I just care about the `,<TAB>`, not the `, <TAB>` (comma and space). The `, <TAB>` completions does not work well with or without this PR. The enum flags case and the `, <TAB>` case might need more investigation and other PRs.

Based on the above, I would like to propose the following fixes:

* Fix test cases: replace `Get-Command` with some commands, for example, `Get-ChildItem`
* Fix PR title represent a smaller problem, for example, "Enable completions for array parameter value just after comma without space even if other parameters follow it"

**Comment by matt9ucci** [comment]
> it causes the tab completion code to throw when you try with Enums and ideally it would never throw

Thank you for your advice. I have fixed it this time for sur... probably:smile:. Please review the code.

**Comment by MartinGC94** [comment]
It looks good now. If you want to also fix it with a space between the comma and cursor you can change line 4128 in CompletionCompleters from `if (!arg.ParameterContainsArgument && arg.Argument.Extent.StartOffset > token.Extent.StartOffset)` to:
```
if (token.Kind == TokenKind.Parameter
|| (token.Extent.StartOffset > arg.Argument.Extent.StartOffset && token.Extent.EndOffset < arg.Argument.Extent.EndOffset))
```

This also fixes the scenario where you move the cursor between a parameter with a colon and a current value like: `Get-Command -CommandType: <Tab> Alias`. If you decide to add this change remember to add tests for it.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs [hunk 1]
```diff
@@ -525,7 +525,7 @@ internal List<CompletionResult> GetResultHelper(CompletionContext completionCont
 
                     case TokenKind.Comma:
                         // Handle array elements such as dir .\cd,<tab> || dir -Path: .\cd,<tab>
-                        if (lastAst is ErrorExpressionAst &&
+                        if ((lastAst is ErrorExpressionAst || lastAst is ArrayLiteralAst) &&
```
[the remark this answers is not in the document]
Comment (matt9ucci) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs]: Fixed it and the next line too.

**Comment by matt9ucci** [comment]
Updated @MartinGC94's suggested lines (thanks!) and test cases.

**Review by daxian-dbw — APPROVED** [review]
LGTM. Made a couple updates on the comments.

**Pull request merged**